azure/services/routetables/client.go

```diff
@@ -54,43 +50,112 @@ func newRouteTablesClient(subscriptionID string, baseURI string, authorizer auto
+	defer cancel()
+
 	err = future.WaitForCompletionRef(ctx, ac.routetables.Client)
 	if err != nil {
-		return err
+		// if an error occurs, return the future.
+		// this means the long-running operation didn't finish in the specified timeout.
+		return nil, &future, err
 	}
-	_, err = future.Result(ac.routetables)
-	return err
+	result, err := future.Result(ac.routetables)
+	// if the operation completed, return a nil future
+	return result, nil, err
 }
 
-// Delete deletes the specified route table.
-func (ac *azureClient) Delete(ctx context.Context, resourceGroupName, rtName string) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "routetables.AzureClient.Delete")
+// DeleteAsync deletes a route table asynchronously. DeleteAsync sends a DELETE
+// request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+func (ac *azureClient) DeleteAsync(ctx context.Context, spec azure.ResourceSpecGetter) (azureautorest.FutureAPI, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "routetables.azureClient.DeleteAsync")
 	defer done()
 
-	future, err := ac.routetables.Delete(ctx, resourceGroupName, rtName)
+	future, err := ac.routetables.Delete(ctx, spec.ResourceGroupName(), spec.ResourceName())
 	if err != nil {
-		return err
+		return nil, err
 	}
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
+	defer cancel()
+
 	err = future.WaitForCompletionRef(ctx, ac.routetables.Client)
 	if err != nil {
-		return err
+		// if an error occurs, return the future.
+		// this means the long-running operation didn't finish in the specified timeout.
+		return &future, err
 	}
 	_, err = future.Result(ac.routetables)
-	return err
+	// if the operation completed, return a nil future.
+	return nil, err
+}
+
+// IsDone returns true if the long-running operation has completed.
+func (ac *azureClient) IsDone(ctx context.Context, future azureautorest.FutureAPI) (bool, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "routetables.azureClient.IsDone")
+	defer done()
+
+	isDone, err := future.DoneWithContext(ctx, ac.routetables)
+	if err != nil {
+		return false, errors.Wrap(err, "failed checking if the operation was complete")
+	}
+
+	return isDone, nil
+}
+
+// Result fetches the result of a long-running operation future.
+func (ac *azureClient) Result(ctx context.Context, futureData azureautorest.FutureAPI, futureType string) (interface{}, error) {
+	_, _, done := tele.StartSpanWithLogger(ctx, "routetables.azureClient.Result")
+	defer done()
+
+	if futureData == nil {
+		return nil, errors.Errorf("cannot get result from nil future")
+	}
+	var result func(client network.RouteTablesClient) (RT network.RouteTable, err error)
+
+	switch futureType {
+	case infrav1.PutFuture:
+		var future *network.RouteTablesCreateOrUpdateFuture
+		jsonData, err := futureData.MarshalJSON()
```

Comment:
Is there a reason for not directly casting `futureData` to `RouteTablesCreateOrUpdateFuture`? I presume `RouteTablesCreateOrUpdateFuture` implements `FutureApi`.

Reply:
Unfortunately this doesn't work right now because the future type is actually azureautorest.Future, not `RouteTablesCreateOrUpdateFuture`, because it was converted back to an Azure Future from the CAPZ infrav1.Future type stored in Status: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/converters/futures.go#L49

Now, we could change the above conversion by either a) adding a separate conversion function for each service, or b) have a switch statement that maps different Azure SDK future type based on the infrav1.Future Type and Name, and this would allow us to do the casting directly. However, that's arguably worse than the few extra lines of code we have to marshal and unmarshal the future above, since the matrix of Future types is so big (one for each type of operation PUT, PATCH, DELETE for each service), we'd have about ~ 25 services * 3 = 75 future types to handle (probably a bit less since not every service implements each future type), and it would be very error-prone.

It's unfortunate that Result() isn't part of the `FutureApi` interface, that would solve the problem, but it's not possible because the Result return is type-specific: https://github.com/Azure/azure-sdk-for-go/issues/14433#issuecomment-802539626

Reply:
this is one of the alternatives: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/scalesets/client.go#L209

Instead of passing in a FutureAPI interface, we directly pass in the infrav1.Future. IMO the above implementation is a bit cleaner (and that's the approach I've adopted in refactors of other services so far) because it doesn't require as tight of a coupling between the client and the API types. Also, we had to jump through some hoops at the time and define our own adapter interfaces to be able to get the Result func from the future: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/47ead2a504fc0165035e9e8c020b54d7422d9b87/azure/services/scalesets/client.go#L59-L71

Reply:
cc @devigned

Reply:
Thanks for the explanation, would it make sense to add a comment that explains this or atleast a link to your comment above?

Reply:
Added a comment there and in the other places that do this, let me know what you think

Reply:
Looks good, Thanks! Maybe in future, we can try to move these pieces of code to some util file, so that they are all in one place.

Reply:
that's what I was thinking too but unfortunately we can't share code because the `RouteTablesCreateOrUpdateFuture` type is unique to the service and to the operation so we'd need a different helper func for each one... At least not until generics come along :)